flocker/node/agents/cinder.py

```diff
@@ -571,43 +571,89 @@ def destroy_volume(self, blockdevice_id):
-            # If the device path does not exist, either virtio driver is
-            # not being used (e.g. Rackspace), or the user has modified
-            # their udev rules.  The following code relies on Cinder
-            # returning the correct device path, which appears to work
-            # for Rackspace and will work with virtio if no disks have
-            # been attached outside Cinder.
-            try:
-                cinder_volume = self.cinder_volume_manager.get(blockdevice_id)
-            except CinderNotFound:
-                raise UnknownVolume(blockdevice_id)
+    def _get_device_path_virtio_blk(self, volume):
+        """
+        The virtio_blk driver allows a serial number to be assigned to virtual blockdevices.
+        OpenStack will set a serial number containing the first 20
+        characters of the Cinder block device ID.
+        The udev daemon will read the serial number and create a
+        symlink to the canonical virtio_blk device path.
+
+        We do this because libvirt does not return the correct device path when additional
+        disks have been attached using a client other than cinder. This is
+        expected behaviour within Cinder and libvirt
+        See https://bugs.launchpad.net/cinder/+bug/1387945 and
+        http://libvirt.org/formatdomain.html#elementsDisks (target section)
+
+        :param volume: The Cinder ``Volume`` which is attached.
+        :returns: ``FilePath`` of the device created by the virtio_blk
+            driver.
+        """
+        expected_path = FilePath(
+            "/dev/disk/by-id/virtio-{}".format(volume.id[:20])
+        )
+        if expected_path.exists():
+            return expected_path.realpath()
+        else:
+            raise UnattachedVolume(volume.id)
+
+    def _get_device_path_other(self, volume):
+        """
+        On other OpenStack hypervisors (Rackspace / Xen) the Cinder
+        API reports the correct device path.
+
+        :param volume: The Cinder ``Volume`` which is attached.
+        :returns: ``FilePath`` of the device created by the virtio_blk
+            driver.
+        """
+        if volume.attachments:
+            attachment = volume.attachments[0]
+            if len(volume.attachments) > 1:
+                # As far as we know you can not have more than one attachment,
+                # but, perhaps we're wrong and there should be a test for the
+                # multiple attachment case.  FLOC-1854.
+                # Log a message if this ever happens.
+                Message.new(
+                    message_type=(
+                        u'flocker:node:agents:blockdevice:openstack:get_device_path:'
+                        u'unexpected_multiple_attachments'
+                    ),
+                    volume_id=unicode(volume.id),
+                    attachment_devices=u','.join(
+                        unicode(a['device']) for a in volume.attachments
+                    ),
+                ).write()
```

Comment:
This isn't really related to this branch, but I noticed that in catching ``ValueError`` we weren't distinguishing between the no attachements and the multiple attachments situations. I'm now logging the multiple attachments situation in case it ever occurs.